PolyEngine/Core/Src/BTreeMap.hpp

```diff
@@ -0,0 +1,532 @@
+					auto remove_result = to_remove.leaf_remove();
+
+					std::swap(key_ref, remove_result.k);
+					std::swap(val_ref, remove_result.v);
+
+					handle_underflow(to_remove.node_ref);
+					return kv{std::move(remove_result.k), std::move(remove_result.v)};
+				}
+			}
+
+			V Remove() { return this->RemoveKV().value; }
+
+		private:
+			template<typename Val>
+			V& or_insert(Val&& value) {
+				if (this->key) {
+					return this->VacantInsert(std::forward<Val>(value));
+				} else {
+					return this->OccupiedGet();
+				}
+			}
+
+			template<typename Val>
+			V& vacant_insert(Val&& value) {
+				ASSERTE(this->IsVacant(), "Cannot insert into already occupied map entry!");
+
+				this->map_len += 1;
+
+				auto insert_result = this->kv_ref.leaf_insert(this->key.TakeValue(), std::forward<Val>(value));
+				if (insert_result.fit) {
+					return *insert_result.value_ref;
+				}
+				//insertion resulted in a node split; need to re-insert the extracted key-value pair
+
+				V* v = insert_result.value_ref; //points to the value we just inserted
+				auto ascension_result = insert_result.split_result.old_node_left.ascend();
+
+				for (;;) {
+					if (ascension_result.succeeded) {
+						insert_result = ascension_result.parent.branch_insert(insert_result.split_result.key, insert_result.split_result.value, insert_result.split_result.new_edge_right);
+						if (insert_result.fit) {
+							return *v;
+						}
+						ascension_result = insert_result.split_result.old_node_left.ascend();
+					} else {
+						Root* root = ascension_result.self.root;
+						root->push_level().push_back(insert_result.split_result.key, insert_result.split_result.value, insert_result.split_result.new_edge_right);
+						return *v;
+					}
+				}
+			}
+
+			template<typename Val>
+			V occupied_insert(Val&& value) {
+				ASSERTE(!this->IsVacant(), "Cannot replace (non-existent) value in a vacant map entry!");
+				//std::swap(value, this->OccupiedGet());
+				//return value;
+				auto old = std::move(this->OccupiedGet());
+				this->OccupiedGet() = std::forward<Val>(value);
+				return old;
+			}
+
+		private:
+			Entry(KVERef kv_ref, size_t& len, const K&  key) : key(key),            kv_ref(kv_ref), map_len(len) {}
+			Entry(KVERef kv_ref, size_t& len,       K&& key) : key(std::move(key)), kv_ref(kv_ref), map_len(len) {}
+			Entry(KVERef kv_ref, size_t& len)                : key{},               kv_ref(kv_ref), map_len(len) {}
+
+		private:
+			Optional<K> key; //todo(vuko): store a reference instead (need to parametrize the class on reference type; lvalue-ref -> opt<T&>, rvalue-ref -> opt<T>)
+			KVERef kv_ref;
+			size_t& map_len;
+		};
+
+		struct KV {
+			const K& key;
+			V& value;
+			const KV* operator->() const { return this; } //note(vuko): needed by the iterator's own operator ->, since it returns KV by value
+		};
+
+		class Iter : public std::iterator<std::bidirectional_iterator_tag, KV> {
```

Comment:
>Rename

Wilco.
>Provide ConstIterator

Oh boy, more boilerplate... I wonder if I should move those associated classes out of line to help the readability.